### ISSUE TITLE
Web UI stuck on "STOPPING"

### ISSUE THREAD

**Issue by the--dud** [locustio/locust#2046]
### Describe the bug
Running locust 2.8.3 on GCP cluster, with docker images python:3.10.2-slim-buster.
I have 1 master and 5 workers. When I press Stop in the web UI, it get's stuck saying STOPPING, and never changes to STOPPED.
This makes it look like the workers are still working, but they are not actually. The workers have stopped and are in ready state.

Logs from master:
```
[2022-03-10 10:21:36,680] locust-master-57bfcc5445-n62pd/INFO/locust.main: Starting web interface at http://0.0.0.0:8089 (accepting connections from all network interfaces)
[2022-03-10 10:21:36,693] locust-master-57bfcc5445-n62pd/INFO/locust.main: Starting Locust 2.8.3
[2022-03-10 10:21:51,492] locust-master-57bfcc5445-n62pd/INFO/locust.runners: Client 'locust-worker-69669f574d-ncvf5_11614c89a3914fd8a4ae11a8365f046d' reported as ready. Currently 1 clients ready to swarm.
[2022-03-10 10:21:52,267] locust-master-57bfcc5445-n62pd/INFO/locust.runners: Client 'locust-worker-69669f574d-tqxqv_70ac83470e4845f495190d98489e30c1' reported as ready. Currently 2 clients ready to swarm.
[2022-03-10 10:21:54,416] locust-master-57bfcc5445-n62pd/INFO/locust.runners: Client 'locust-worker-69669f574d-m9fc5_912176e2721949ab95871d8c9f29977b' reported as ready. Currently 3 clients ready to swarm.
[2022-03-10 10:21:55,716] locust-master-57bfcc5445-n62pd/INFO/locust.runners: Client 'locust-worker-69669f574d-4fltb_4bfda0642dbe4463bd7c1bc3bda676d8' reported as ready. Currently 4 clients ready to swarm.
[2022-03-10 10:21:57,329] locust-master-57bfcc5445-n62pd/INFO/locust.runners: Client 'locust-worker-69669f574d-775nk_7c6fa568b8d2456399d05edfdb171152' reported as ready. Currently 5 clients ready to swarm.
[2022-03-10 10:22:20,491] locust-master-57bfcc5445-n62pd/INFO/locust.runners: Sending spawn jobs of 10 users at 10.00 spawn rate to 5 ready clients
[2022-03-10 10:22:20,575] locust-master-57bfcc5445-n62pd/INFO/locust.runners: All users spawned: {"AuthUser": 10} (10 total users)
[2022-03-10 10:22:41,006] locust-master-57bfcc5445-n62pd/INFO/locust.runners: Removing locust-worker-69669f574d-m9fc5_912176e2721949ab95871d8c9f29977b client from running clients
[2022-03-10 10:22:41,006] locust-master-57bfcc5445-n62pd/INFO/locust.runners: Client 'locust-worker-69669f574d-m9fc5_912176e2721949ab95871d8c9f29977b' reported as ready. Currently 5 clients ready to swarm.
[2022-03-10 10:22:41,006] locust-master-57bfcc5445-n62pd/INFO/locust.runners: Removing locust-worker-69669f574d-ncvf5_11614c89a3914fd8a4ae11a8365f046d client from running clients
[2022-03-10 10:22:41,007] locust-master-57bfcc5445-n62pd/INFO/locust.runners: Client 'locust-worker-69669f574d-ncvf5_11614c89a3914fd8a4ae11a8365f046d' reported as ready. Currently 5 clients ready to swarm.
[2022-03-10 10:22:41,007] locust-master-57bfcc5445-n62pd/INFO/locust.runners: Removing locust-worker-69669f574d-775nk_7c6fa568b8d2456399d05edfdb171152 client from running clients
[2022-03-10 10:22:41,007] locust-master-57bfcc5445-n62pd/INFO/locust.runners: Client 'locust-worker-69669f574d-775nk_7c6fa568b8d2456399d05edfdb171152' reported as ready. Currently 5 clients ready to swarm.
[2022-03-10 10:22:41,007] locust-master-57bfcc5445-n62pd/INFO/locust.runners: Removing locust-worker-69669f574d-tqxqv_70ac83470e4845f495190d98489e30c1 client from running clients
[2022-03-10 10:22:41,007] locust-master-57bfcc5445-n62pd/INFO/locust.runners: Removing locust-worker-69669f574d-4fltb_4bfda0642dbe4463bd7c1bc3bda676d8 client from running clients
[2022-03-10 10:22:41,007] locust-master-57bfcc5445-n62pd/INFO/locust.runners: Client 'locust-worker-69669f574d-4fltb_4bfda0642dbe4463bd7c1bc3bda676d8' reported as ready. Currently 4 clients ready to swarm.
[2022-03-10 10:22:41,007] locust-master-57bfcc5445-n62pd/INFO/locust.runners: Client 'locust-worker-69669f574d-tqxqv_70ac83470e4845f495190d98489e30c1' reported as ready. Currently 5 clients ready to swarm.
```

Response from https://auth-locust.signicat.dev/stats/requests
```
{
  "current_response_time_percentile_50": null, 
  "current_response_time_percentile_95": null, 
  "errors": [], 
  "fail_ratio": 0.0, 
  "state": "stopping", 
  "stats": [
    {
      "avg_content_length": 2635.0, 
      "avg_response_time": 118.18601360375231, 
      "current_fail_per_sec": 0.0, 
      "current_rps": 2.8, 
      "max_response_time": 297.0, 
      "median_response_time": 100, 
      "method": "GET", 
      "min_response_time": 8.0, 
      "name": HIDDEN, 
      "ninetieth_response_time": 200.0, 
      "ninety_ninth_response_time": 300.0, 
      "num_failures": 0, 
      "num_requests": 55, 
      "safe_name":HIDDEN
    }, 
    {
      "avg_content_length": 922.0, 
      "avg_response_time": 171.46125374446538, 
      "current_fail_per_sec": 0.0, 
      "current_rps": 3.1, 
      "max_response_time": 401.0, 
      "median_response_time": 190.0, 
      "method": "GET", 
      "min_response_time": 13.0, 
      "name": HIDDEN, 
      "ninetieth_response_time": 290.0, 
      "ninety_ninth_response_time": 400.0, 
      "num_failures": 0, 
      "num_requests": 55, 
      "safe_name": HIDDEN
    }, 
    {
      "avg_content_length": 8432.0, 
      "avg_response_time": 574.5072794367848, 
      "current_fail_per_sec": 0.0, 
      "current_rps": 10.9, 
      "max_response_time": 5829.0, 
      "median_response_time": 410.0, 
      "method": "GET", 
      "min_response_time": 119.0, 
      "name": HIDDEN, 
      "ninetieth_response_time": 610.0, 
      "ninety_ninth_response_time": 5000.0, 
      "num_failures": 0, 
      "num_requests": 197, 
      "safe_name": HIDDEN
    }, 
    {
      "avg_content_length": 898.0, 
      "avg_response_time": 164.27931896370345, 
      "current_fail_per_sec": 0.0, 
      "current_rps": 27.6, 
      "max_response_time": 459.0, 
      "median_response_time": 180.0, 
      "method": "POST", 
      "min_response_time": 8.0, 
      "name": HIDDEN
      "ninetieth_response_time": 280.0, 
      "ninety_ninth_response_time": 390.0, 
      "num_failures": 0, 
      "num_requests": 448, 
      "safe_name": HIDDEN
    }, 
    {
      "avg_content_length": 2992.1099337748346, 
      "avg_response_time": 268.4843293363414, 
      "current_fail_per_sec": 0.0, 
      "current_rps": 44.4, 
      "max_response_time": 5829.0, 
      "median_response_time": 200.0, 
      "method": null, 
      "min_response_time": 8.0, 
      "name": "Aggregated", 
      "ninetieth_response_time": 420.0, 
      "ninety_ninth_response_time": 2600.0, 
      "num_failures": 0, 
      "num_requests": 755, 
      "safe_name": "Aggregated"
    }
  ], 
  "total_rps": 44.4, 
  "user_count": 0, 
  "workers": [
    {
      "cpu_usage": 0.0, 
      "id": "locust-worker-69669f574d-m9fc5_912176e2721949ab95871d8c9f29977b", 
      "memory_usage": 52801536, 
      "state": "ready", 
      "user_count": 0
    }, 
    {
      "cpu_usage": 0.0, 
      "id": "locust-worker-69669f574d-ncvf5_11614c89a3914fd8a4ae11a8365f046d", 
      "memory_usage": 52969472, 
      "state": "ready", 
      "user_count": 0
    }, 
    {
      "cpu_usage": 0.2, 
      "id": "locust-worker-69669f574d-775nk_7c6fa568b8d2456399d05edfdb171152", 
      "memory_usage": 53260288, 
      "state": "ready", 
      "user_count": 0
    }, 
    {
      "cpu_usage": 0.0, 
      "id": "locust-worker-69669f574d-4fltb_4bfda0642dbe4463bd7c1bc3bda676d8", 
      "memory_usage": 53157888, 
      "state": "ready", 
      "user_count": 0
    }, 
    {
      "cpu_usage": 0.2, 
      "id": "locust-worker-69669f574d-tqxqv_70ac83470e4845f495190d98489e30c1", 
      "memory_usage": 52875264, 
      "state": "ready", 
      "user_count": 0
    }
  ]
}
```

### Expected behavior
After saying STOPPING for a while, I expect it to say STOPPED.

### Actual behavior
Never says STOPPED, only STOPPING.

### Steps to reproduce

1. Setup 1 master and 5 workers
2. Open web UI
3. Start test
4. Let it run for a while
5. Stop test
6. Observe that status is stuck on STOPPING and never goes to STOPPED.

### Environment

- OS: python:3.10.2-slim-buster
- Python version: python:3.10.2-slim-buster
- Locust version: (please dont file issues for anything but the most recent release or prerelease builds)
- Locust command line that you ran:
- Locust file contents (anonymized if necessary):

**Comment by the--dud** [comment]
![image](https://user-images.githubusercontent.com/913780/157644249-f10f5944-374d-49e3-b990-0a100cb1e72e.png)

**Comment by kurtmckee** [comment]
I'm seeing output that may be related when running the unit tests. This is specifically happening with Python 3.7. Steps I followed:

```shell
# Clone the repo and create a virtual environment.
# The virtual environment is in '.hg/' to hide from flake8.
git clone git@github.com:locustio/locust.git locus2
cd locus2
python -m venv .hg
. .hg/bin/activate

# Update pip, et al, and install tox.
python -m pip install --upgrade pip setuptools wheel
python -m pip install tox

# Run the tests.
tox
```

The errors are inconsistent, but I have seen one of the following two errors on Python 3.6, 3.7, and 3.10:

```
======================================================================
FAIL: test_distributed_shape (locust.test.test_runners.TestMasterWorkerRunners)
Full integration test that starts both a MasterRunner and three WorkerRunner instances
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/kurt/dev/locust/locust/test/test_runners.py", line 1112, in test_distributed_shape
    self.assertEqual("stopped", master.state)
AssertionError: 'stopped' != 'stopping'
- stopped
+ stopping
```

```
======================================================================
FAIL: test_distributed_shape_with_fixed_users (locust.test.test_runners.TestMasterWorkerRunners)
Full integration test that starts both a MasterRunner and three WorkerRunner instances
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/kurt/dev/locus2/locust/test/test_runners.py", line 1208, in test_distributed_shape_with_fixed_users
    self.assertEqual(STATE_STOPPED, master.state)
AssertionError: 'stopped' != 'stopping'
- stopped
+ stopping
```

I don't know if this is related. @cyberw Please let me know if this should be posted to a separate issue.

**Comment by cyberw** [comment]
It is entirely possible that they are related - thanks for pointing that out. I dont have the energy/time to investigate at the moment (thanks post-covid..) but would be very happy if someone would have a look at it (even just getting a stable repro)

**Comment by kurtmckee** [comment]
Hmm...might not be related.

https://github.com/locustio/locust/blob/68829f071033e8cef38e24e3a2fcd89419eeb75a/locust/test/test_runners.py#L1103-L1112

I isolated just this test and found that I could reproduce intermittent failures by reducing that `time.sleep(2)` line down to `1.1`. If I set it to `1` then it just always failed. It's likely that this is just a timing issue with that `time.sleep()` call in the context of the larger test framework.

What I'm seeing is probably unrelated to this bug report. I hope you feel better soon!
